File: backend/app/scoring.py
"""Scoring math (Section 3).

The frontend re-implements this for instant re-ranking when weights change; this
server copy backs CSV export and any server-side sorting. Keep the two in sync —
both derive ``$/sqft``, the commute rating, and the weighted/combined scores rather
than storing them.
"""
from __future__ import annotations

from typing import Optional

from .models import DIMENSIONS


def commute_minutes(listing: dict) -> Optional[float]:
    """Manual override wins (Section 8 fallback), else the cached directions result."""
    if listing.get("commute_minutes_manual") is not None:
        return listing["commute_minutes_manual"]
    commute = listing.get("commute") or {}
    return commute.get("minutes")


def minutes_to_rating(minutes: Optional[float], bands: list[dict]) -> Optional[int]:
    """Map commute minutes onto a 1-5 rating via the editable bands. Bands are
    evaluated ascending; a band with ``max_minutes is None`` is the catch-all."""
    if minutes is None:
        return None
    ordered = sorted(bands, key=lambda b: (b["max_minutes"] is None, b["max_minutes"] or 0))
    for band in ordered:
        if band["max_minutes"] is None or minutes <= band["max_minutes"]:
            return band["rating"]
    return ordered[-1]["rating"] if ordered else None


def effective_ratings(listing: dict, rater: str, settings: dict) -> dict[str, int]:
    """A rater's ratings with the commute dimension filled in from minutes when the
    human hasn't overridden it. Only dimensions with a numeric value are returned."""
    raw = (listing.get("ratings") or {}).get(rater) or {}
    out: dict[str, int] = {d: v for d, v in raw.items() if isinstance(v, (int, float)) and v is not None}
    if out.get("commute") is None:
        derived = minutes_to_rating(commute_minutes(listing), settings["commute_bands"])
        if derived is not None:
            out["commute"] = derived
    return out


def weighted_score(ratings: dict[str, int], weights: dict[str, float]) -> Optional[float]:
    """Weighted average over the dimensions that have a value, renormalizing the
    weights across only those present so a missing rating doesn't drag the score to
    zero. Returns a 1-5 float, or None if nothing is rated."""
    num = 0.0
    denom = 0.0
    for dim in DIMENSIONS:
        val = ratings.get(dim)
        if val is None:
            continue
        w = weights.get(dim, 0.0)
        num += val * w
        denom += w
    if denom == 0:
        return None
    return round(num / denom, 3)


def rater_score(listing: dict, rater: str, settings: dict) -> Optional[float]:
    return weighted_score(effective_ratings(listing, rater, settings), settings["weights"])


def combined_score(listing: dict, settings: dict) -> Optional[float]:
    """Mean of whichever humans have scored. Used as the default ranking key. Falls
    back to the model's ``auto`` rating when neither human has rated, so scraped-but-
    unrated listings still rank instead of dropping to the bottom. Human ratings
    always win — ``auto`` is consulted only when both humans are silent."""
    scores = [s for r in ("you", "fiance") if (s := rater_score(listing, r, settings)) is not None]
    if scores:
        return round(sum(scores) / len(scores), 3)
    return rater_score(listing, "auto", settings)


def ranked_by(listing: dict) -> str:
    """Who produced the combined score: a human rated it, or the model auto-scored it."""
    human = (listing.get("ratings") or {})
    if any((human.get(r) or {}).get(d) is not None for r in ("you", "fiance") for d in DIMENSIONS):
        return "human"
    if (listing.get("ratings") or {}).get("auto"):
        return "auto"
    return "none"


def dollar_per_sqft(listing: dict) -> Optional[float]:
    rent, sqft = listing.get("rent"), listing.get("sqft")
    if not rent or not sqft:
        return None
    return round(rent / sqft, 2)


def divergence(listing: dict, settings: dict) -> dict:
    """Per-dimension disagreement between the two raters. Flagged when any dimension
    differs by >= the configured threshold (Section 7, default 2)."""
    threshold = settings.get("divergence_threshold", 2)
    you = (listing.get("ratings") or {}).get("you") or {}
    fiance = (listing.get("ratings") or {}).get("fiance") or {}
    rows = []
    worst = 0
    for dim in DIMENSIONS:
        a, b = you.get(dim), fiance.get(dim)
        if a is None or b is None:
            continue
        diff = abs(a - b)
        worst = max(worst, diff)
        rows.append({"dimension": dim, "you": a, "fiance": b, "diff": diff})
    return {"flagged": worst >= threshold, "max_diff": worst, "dimensions": rows}


def enrich(listing: dict, settings: dict) -> dict:
    """Attach all derived fields to a listing for API responses. Stored data is left
    untouched; these live only on the wire."""
    return {
        **listing,
        "derived": {
            "dollar_per_sqft": dollar_per_sqft(listing),
            "commute_minutes": commute_minutes(listing),
            "score_you": rater_score(listing, "you", settings),
            "score_fiance": rater_score(listing, "fiance", settings),
            "score_auto": rater_score(listing, "auto", settings),
            "score_combined": combined_score(listing, settings),
            "ranked_by": ranked_by(listing),
            "divergence": divergence(listing, settings),
        },
    }

File: frontend/src/scoring.ts
// Client-side mirror of backend scoring (app/scoring.py) so the ranking re-computes
// instantly when weights or ratings change — no round-trip. Keep in sync with the
// server copy. Also holds the faceted filter logic used by the filter rail.

import {
  DIMENSIONS,
  type CommuteBand,
  type Derived,
  type Dimension,
  type Divergence,
  type FilterQuery,
  type Listing,
  type RatingsKey,
  type Settings,
} from "./types";

export function commuteMinutes(l: Listing): number | null {
  if (l.commute_minutes_manual != null) return l.commute_minutes_manual;
  return l.commute?.minutes ?? null;
}

export function minutesToRating(minutes: number | null, bands: CommuteBand[]): number | null {
  if (minutes == null) return null;
  const ordered = [...bands].sort((a, b) => {
    const am = a.max_minutes ?? Infinity;
    const bm = b.max_minutes ?? Infinity;
    return am - bm;
  });
  for (const band of ordered) {
    if (band.max_minutes == null || minutes <= band.max_minutes) return band.rating;
  }
  return ordered.length ? ordered[ordered.length - 1].rating : null;
}

function effectiveRatings(l: Listing, rater: RatingsKey, s: Settings): Partial<Record<Dimension, number>> {
  const raw = l.ratings?.[rater] ?? {};
  const out: Partial<Record<Dimension, number>> = {};
  for (const d of DIMENSIONS) {
    const v = raw[d];
    if (typeof v === "number") out[d] = v;
  }
  if (out.commute == null) {
    const derived = minutesToRating(commuteMinutes(l), s.commute_bands);
    if (derived != null) out.commute = derived;
  }
  return out;
}

export function weightedScore(
  ratings: Partial<Record<Dimension, number>>,
  weights: Record<Dimension, number>,
): number | null {
  let num = 0;
  let denom = 0;
  for (const d of DIMENSIONS) {
    const v = ratings[d];
    if (v == null) continue;
    const w = weights[d] ?? 0;
    num += v * w;
    denom += w;
  }
  if (denom === 0) return null;
  return round(num / denom, 3);
}

export function raterScore(l: Listing, rater: RatingsKey, s: Settings): number | null {
  return weightedScore(effectiveRatings(l, rater, s), s.weights);
}

// Falls back to the model's "auto" rating when neither human has scored, so scraped
// listings still rank. Human ratings always win. Mirrors scoring.py combined_score.
export function combinedScore(l: Listing, s: Settings): number | null {
  const scores = (["you", "fiance"] as RatingsKey[])
    .map((r) => raterScore(l, r, s))
    .filter((x): x is number => x != null);
  if (scores.length) return round(scores.reduce((a, b) => a + b, 0) / scores.length, 3);
  return raterScore(l, "auto", s);
}

// Whether the combined score is a human rating or the model's auto fallback.
export function rankedBy(l: Listing): "human" | "auto" | "none" {
  const human = l.ratings ?? {};
  const hasHuman = (["you", "fiance"] as const).some((r) =>
    DIMENSIONS.some((d) => human[r]?.[d] != null),
  );
  if (hasHuman) return "human";
  return l.ratings?.auto ? "auto" : "none";
}

export function dollarPerSqft(l: Listing): number | null {
  if (!l.rent || !l.sqft) return null;
  return round(l.rent / l.sqft, 2);
}

export function divergence(l: Listing, s: Settings): Divergence {
  const threshold = s.divergence_threshold ?? 2;
  const you = l.ratings?.you ?? {};
  const fiance = l.ratings?.fiance ?? {};
  const rows: Divergence["dimensions"] = [];
  let worst = 0;
  for (const d of DIMENSIONS) {
    const a = you[d];
    const b = fiance[d];
    if (a == null || b == null) continue;
    const diff = Math.abs(a - b);
    worst = Math.max(worst, diff);
    rows.push({ dimension: d, you: a, fiance: b, diff });
  }
  return { flagged: worst >= threshold, max_diff: worst, dimensions: rows };
}

export function derive(l: Listing, s: Settings): Derived {
  return {
    dollar_per_sqft: dollarPerSqft(l),
    commute_minutes: commuteMinutes(l),
    score_you: raterScore(l, "you", s),
    score_fiance: raterScore(l, "fiance", s),
    score_auto: raterScore(l, "auto", s),
    score_combined: combinedScore(l, s),
    ranked_by: rankedBy(l),
    divergence: divergence(l, s),
  };
}

// --- filtering (mirrors backend _passes) -----------------------------------

export function passes(
  l: Listing,
  s: Settings,
  f: FilterQuery,
  catMap: Record<string, string>,
): boolean {
  const d = derive(l, s);
  // Hide noise by default: dead (expired/deleted) links and marked duplicates.
  if (!f.show_expired && (l.link_status === "dead" || l.duplicate_of)) return false;
  if (f.rent_min != null && (l.rent == null || l.rent < f.rent_min)) return false;
  if (f.rent_max != null && (l.rent == null || l.rent > f.rent_max)) return false;
  if (f.sqft_min != null && (l.sqft == null || l.sqft < f.sqft_min)) return false;
  if (f.beds != null && (l.beds == null || l.beds < f.beds)) return false;
  if (f.commute_max != null && (d.commute_minutes == null || d.commute_minutes > f.commute_max))
    return false;
  if (f.min_score != null && (d.score_combined == null || d.score_combined < f.min_score))
    return false;
  if (f.neighborhoods?.length && !f.neighborhoods.includes(l.neighborhood)) return false;
  if (f.status?.length && !f.status.includes(l.status)) return false;

  const amen = l.amenities ?? {};
  if (f.in_unit_laundry && !amen.in_unit_laundry) return false;
  if (f.parking && !amen.parking) return false;
  if (f.pet_friendly) {
    const policy = (amen.pet_policy ?? "").toLowerCase();
    if (!policy || policy.includes("no pet")) return false;
  }

  // Faceted tags: AND across categories, OR within a category.
  if (f.tags?.length) {
    const wantByCat: Record<string, Set<string>> = {};
    for (const label of f.tags) {
      const cat = catMap[label] ?? "custom";
      (wantByCat[cat] ??= new Set()).add(label);
    }
    const have = new Set(l.tags ?? []);
    for (const wanted of Object.values(wantByCat)) {
      if (![...wanted].some((t) => have.has(t))) return false;
    }
  }
  return true;
}

// Faceted counts: for each tag, how many listings would match if that tag were
// added to the current filter (classic faceting — the tag's own category is
// relaxed so toggling it on doesn't zero out its siblings).
export function facetCounts(
  listings: Listing[],
  s: Settings,
  f: FilterQuery,
  catMap: Record<string, string>,
  allTags: string[],
): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const label of allTags) {
    const cat = catMap[label] ?? "custom";
    // Drop other tags in the same category from the active filter, then require this one.
    const otherCats = (f.tags ?? []).filter((t) => (catMap[t] ?? "custom") !== cat);
    const probe: FilterQuery = { ...f, tags: [...otherCats, label] };
    counts[label] = listings.filter((l) => passes(l, s, probe, catMap)).length;
  }
  return counts;
}

export type SortKey = "score" | "rent" | "ppsqft" | "sqft" | "commute" | "recency";

export function sortListings(
  listings: Listing[],
  s: Settings,
  key: SortKey,
  dir: "asc" | "desc",
): Listing[] {
  const val = (l: Listing): [boolean, number | string] => {
    const d = derive(l, s);
    switch (key) {
      case "rent":
        return [l.rent == null, l.rent ?? 0];
      case "ppsqft":
        return [d.dollar_per_sqft == null, d.dollar_per_sqft ?? 0];
      case "sqft":
        return [l.sqft == null, l.sqft ?? 0];
      case "commute":
        return [d.commute_minutes == null, d.commute_minutes ?? 0];
      case "recency":
        return [false, l.created_at];
      default:
        return [d.score_combined == null, d.score_combined ?? 0];
    }
  };
  return [...listings].sort((a, b) => {
    const [an, av] = val(a);
    const [bn, bv] = val(b);
    if (an !== bn) return an ? 1 : -1; // nulls always last
    const cmp = av < bv ? -1 : av > bv ? 1 : 0;
    return dir === "desc" ? -cmp : cmp;
  });
}

function round(n: number, dp: number): number {
  const f = 10 ** dp;
  return Math.round(n * f) / f;
}
